net/portmapper/portmapper.go

```diff
@@ -0,0 +1,601 @@
+}
+
+func (c *Client) sawPCPRecently() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.pcpSawTime.After(time.Now().Add(-trustServiceStillAvailableDuration))
+}
+
+func (c *Client) sawUPnPRecently() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.uPnPSawTime.After(time.Now().Add(-trustServiceStillAvailableDuration))
+}
+
+func closeCloserOnContextDone(ctx context.Context, c io.Closer) (stop func()) {
+	// Close uc on ctx being done.
+	ctxDone := ctx.Done()
+	if ctxDone == nil {
+		return func() {}
+	}
+	stopWaitDone := make(chan struct{})
+	go func() {
+		select {
+		case <-stopWaitDone:
+		case <-ctxDone:
+			c.Close()
+		}
+	}()
+	return func() { close(stopWaitDone) }
+}
+
+// NoMappingError is returned by CreateOrGetMapping when no NAT
+// mapping could be returned.
+type NoMappingError struct {
+	Err error
+}
+
+func (nme NoMappingError) Unwrap() error { return nme.Err }
+func (nme NoMappingError) Error() string { return fmt.Sprintf("no NAT mapping available: %v", nme.Err) }
+
+func IsNoMappingError(err error) bool {
+	_, ok := err.(NoMappingError)
+	return ok
+}
+
+var (
+	ErrNoPortMappingServices = errors.New("no port mapping services were found")
+	ErrGatewayDetected       = errors.New("failed to look gateway address")
+)
+
+// CreateOrGetMapping either creates a new mapping or returns a cached
+// valid one.
+//
+// If no mapping is available, the error will be of type
+// NoMappingError; see IsNoMappingError.
+func (c *Client) CreateOrGetMapping(ctx context.Context) (external netaddr.IPPort, err error) {
+	gw, myIP, ok := c.gatewayAndSelfIP()
+	if !ok {
+		return netaddr.IPPort{}, NoMappingError{ErrGatewayDetected}
+	}
+
+	c.mu.Lock()
+	localPort := c.localPort
+	m := &pmpMapping{
+		gw:       gw,
+		internal: netaddr.IPPort{IP: myIP, Port: localPort},
+	}
+
+	prevPort := uint16(0) // or 0 if unknown
+
+	// Do we have an existing mapping that's valid?
+	now := time.Now()
+	if m := c.pmpMapping; m != nil {
+		if now.Before(m.useUntil) {
+			defer c.mu.Unlock()
+			return m.external, nil
+		}
+		// The mapping might still be valid, so just try to renew it.
+		prevPort = m.external.Port
+	}
+
+	// If we just did a Probe (e.g. via netchecker) but didn't
+	// find a PMP service, bail out early rather than probing
+	// again. Cuts down latency for most clients.
+	haveRecentPMP := c.sawPMPRecentlyLocked()
+	if haveRecentPMP {
+		m.external.IP = c.pmpPubIP
+	}
+	if c.lastProbe.After(now.Add(-5*time.Second)) && !haveRecentPMP {
+		return netaddr.IPPort{}, NoMappingError{ErrNoPortMappingServices}
+	}
+
+	c.mu.Unlock()
+
+	uc, err := netns.Listener().ListenPacket(ctx, "udp4", ":0")
+	if err != nil {
+		return netaddr.IPPort{}, err
+	}
+	defer uc.Close()
+
+	uc.SetReadDeadline(time.Now().Add(portMapServiceTimeout))
+	defer closeCloserOnContextDone(ctx, uc)()
+
+	pmpAddr := netaddr.IPPort{IP: gw, Port: pmpPort}
+	pmpAddru := pmpAddr.UDPAddr()
+
+	// Ask for our external address if needed.
```

Comment:
Wow, annoying that the port mapping request doesn't include the external IP in its response. I think this is one place where PCP improved significantly, iirc.